### PR TITLE
refactor: job queue based around multi-producer single-consumer channel

### PR DESCRIPTION
**Pull request by WiredSound**
https://app.clickup.com/t/21pccn4

### CODE REVIEW

File: api/src/deployment.rs [hunk 1]
```diff
@@ -365,8 +364,7 @@ impl DeploymentSystem {
     async fn initialise_from_fs(fs_root: &Path) -> Deployments {
         let mut deployments = HashMap::default();
         for project_dir in std::fs::read_dir(fs_root)
-            .unwrap() // safety: api can read the fs root dir
-            .into_iter() {
+            .unwrap() {
```
Comment (rimutaka) [inline api/src/deployment.rs:367]: could still be not safe - fs root probably exists, but it doesn't mean the app can read it

File: lib/src/project.rs [hunk 2]
```diff
@@ -29,9 +29,9 @@ fn deserialize_project_name<'de, D>(deserializer: D) -> Result<ProjectName, D::E
 impl ProjectName {
     pub fn is_valid(hostname: &str) -> bool {
         fn is_valid_char(byte: u8) -> bool {
-            (byte >= b'a' && byte <= b'z')
-                || (byte >= b'A' && byte <= b'Z')
-                || (byte >= b'0' && byte <= b'9')
+            (b'a'..=b'z').contains(&byte)
```
Comment (rimutaka) [inline lib/src/project.rs:32]: I wonder why. This is definitely more elegant, but is probably slower and takes up more memory.

File: api/src/auth.rs [hunk 3]
```diff
@@ -99,7 +99,7 @@ impl UserDirectory {
     fn from_user_file() -> Self {
         let file_path = Self::users_toml_file_path();
         let file_contents: String = std::fs::read_to_string(&file_path)
-            .expect(&format!("this should blow up if the users.toml file is not present at {:?}", &file_path));
+            .unwrap_or_else(|_| panic!("this should blow up if the users.toml file is not present at {:?}", &file_path));
```
Comment (rimutaka) [inline api/src/auth.rs:102]: I'm curious to why replace expect with this construct.

File: lib/src/project.rs (hunk 2 above)
Comment (WiredSound) [reply lib/src/project.rs:32]: I believe this was just a correction of [this Clippy lint.](https://rust-lang.github.io/rust-clippy/master/#manual_range_contains) You raise an interesting point about efficiency - I wonder if the compiler is clever enough to optimise it to be equivalent to the former under the hood or not.

File: api/src/auth.rs (hunk 3 above)
Comment (WiredSound) [reply api/src/auth.rs:102]: Ahh this is again [a Clippy lint](https://rust-lang.github.io/rust-clippy/master/#expect_fun_call) :) Using `unwrap_or_else` avoids the unnecessary call to `format!` should `read_to_string` return `Ok`.

**Pull request merged**